### ISSUE TITLE
[Question] Tree Structure using SourceList

### ISSUE THREAD

**Issue by Skm1221** [reactivemarbles/DynamicData#210]
Hello, 

I try to make tree structure by using SourceList in WPF with ReactiveUI

Since children node order can be changed by drag and drop event, I can't use SourceCache. 

Each node has child node list and one readonlyObservableCollection(`rows`).

`rows` is consist of self and child node's `rows`. 

and when any child node's `rows` is updated, I want to update the parent node's `rows`. 

```
public abstract class TreeListViewNodeViewModel : ReactiveObject, ITreeListViewNodeViewModel
    {
        private readonly ISourceList<ITreeListViewNodeViewModel> childrenSourceList;

        private readonly ReadOnlyObservableCollection<ITreeListViewNodeViewModel> children;

        private readonly ReadOnlyObservableCollection<ITreeListViewNodeViewModel> rows;

        public TreeListViewNodeViewModel(
            IEnumerable<ITreeListViewNodeViewModel> children,
            bool isRoot = false)
        {
            this.childrenSourceList = new SourceList<ITreeListViewNodeViewModel>();
            this.RowSourceList = new SourceList<ITreeListViewNodeViewModel>();
            this.IsRoot = isRoot;
            this.IsExpanded = this.IsRoot;

            if (!this.IsRoot)
            {
                this.RowSourceList.Add(this);
            }

            if (children != null)
            {
                this.AddChildren(children);
            }

            this.RowSourceList
                .Connect()
                .Bind(out this.rows)
                .Subscribe();

            var sharedChildrenSourceList = this.childrenSourceList
                .Connect()
                .Publish();

            sharedChildrenSourceList
                .Bind(out this.children)
                .Subscribe();

            sharedChildrenSourceList
                .WhereReasonsAre(ListChangeReason.Remove, ListChangeReason.RemoveRange)
                .Transform(child =>
                {
                    child.Parent = null;
                    return child;
                })
                .Subscribe();

            sharedChildrenSourceList
                .WhereReasonsAre(ListChangeReason.Add, ListChangeReason.AddRange)
                .Transform(child =>
                {
                    child.Parent = this;
                    return child;
                })
                .Subscribe();

            var changed = sharedChildrenSourceList.MergeMany(node => node.RowSourceList.Connect());

            sharedChildrenSourceList
                .AutoRefreshOnObservable(p => changed)
                .Bind(out this.rows)
                .Subscribe();

            this.WhenAnyValue(x => x.IsExpanded)
                .Do(isExpanded =>
                {
                    if (isExpanded)
                    {
                        this.RowSourceList.AddRange(this.children.SelectMany(child => child.Rows));
                    }
                    else
                    {
                        if (this.RowSourceList.Count > 1)
                        {
                            this.RowSourceList.RemoveRange(1, this.RowSourceList.Count - 1);
                        }
                    }
                })
                .Subscribe();
            sharedChildrenSourceList.Connect();
        }
```

Maybe I think key point is 
```
            var changed = sharedChildrenSourceList.MergeMany(node => node.RowSourceList.Connect());

            sharedChildrenSourceList
                .AutoRefreshOnObservable(p => changed)
                .Bind(out this.rows)
                .Subscribe();
```

But maybe above `.Bind(out this.rows)` override RowSourceList's bind. 

Is any other approach ?

**Comment by RolandPheasant** [comment]
Have you got some unit tests to check?

**Comment by Skm1221** [comment]
Not yet, I just do debug with breakpoint.
I will make it today.

**Comment by Skm1221** [comment]
Hello Roland, 

I made a simple test case 

```
public class TreeListNodeViewModelTest : TreeListViewNodeViewModel
        {
            public TreeListNodeViewModelTest(IEnumerable<ITreeListViewNodeViewModel> children, bool isRoot = false)
                : base(children, isRoot)
            {
            }
        }

[Fact]
        public void Test()
        {
            var rootNode = new TreeListNodeViewModelTest(null, true);

            Assert.Empty(rootNode.Rows);
            Assert.True(rootNode.IsExpanded);

            var childNode1 = new TreeListNodeViewModelTest(null, false);

            Assert.Single(childNode1.Rows);
            Assert.False(childNode1.IsExpanded);

            rootNode.AddChild(childNode1);

            Assert.Single(rootNode.Rows);
            var childNode2 = new TreeListNodeViewModelTest(null, false);
            Assert.Single(childNode2.Rows);
            Assert.False(childNode2.IsExpanded);

            rootNode.AddChild(childNode2);

            Assert.Equal(2, rootNode.Rows.Count());

            var childNodeOfChildNode1 = new TreeListNodeViewModelTest(null, false);
            childNode1.AddChild(childNodeOfChildNode1);

            Assert.Single(childNode1.Rows);
            Assert.Equal(2, rootNode.Rows.Count());

            childNode1.IsExpanded = true;

            Assert.Equal(2, childNode1.Rows.Count());
            Assert.Equal(3, rootNode.Rows.Count());
            Assert.Equal(childNode1, rootNode.Rows.ToList()[0]);
            Assert.Equal(childNodeOfChildNode1, rootNode.Rows.ToList()[1]);
            Assert.Equal(childNode2, rootNode.Rows.ToList()[2]);

            childNode1.IsExpanded = false;

            Assert.Single(childNode1.Rows);
            Assert.Equal(2, rootNode.Rows.Count());
            Assert.Equal(childNode1, rootNode.Rows.ToList()[0]);
            Assert.Equal(childNode2, rootNode.Rows.ToList()[1]);
        }
    }
``` 
```
public void AddChild(ITreeListViewNodeViewModel node)
        {
            this.childrenSourceList.Add(node);
            this.RowSourceList.AddRange(node.Rows);
        }
```

Based on children's `rows`, parent's `rows` have to be changed.

**Comment by Skm1221** [comment]
How can I combine multiple readOnlyObservableCollection to One readOnlyObservableCollection.

I changed my code little bit. 

https://gist.github.com/Skm1221/7faf1751fafb677cce4a205ee4b201d2

Keypoint is I made `ISourceList<ReadOnlyObservableCollection<ITreeListViewNodeViewModel>> childrenRowsSourceList`

From `childrenRowsSourceList`, I made one combined collection.
```
this.childrenRowsSourceList
                .Connect()
                .TransformMany(x => x)
                .Bind(out this.rows)
                .DisposeMany()
                .Subscribe();
```

but, when some collection in `childrenRowSourceList` updated, updated item's position is wrong.
![image](https://user-images.githubusercontent.com/12227644/53679832-7949d480-3d15-11e9-8b13-232c6d2d34f4.png)

when I click expand button, operation raw position have to be second, but third. 

when I check childrenRowSourceList, 
Items are [Collection('folder', 'operation'), Collection('folder)]. Maybe When I use transformMany, some operator may need. 

So I added `AutoRefresh()` after `transformMany`, then updated collection's position is changed to end position.
